File: Send-Email-SMS-Azure-Function-MessageFlow-Java/MessageFlowSmsFunction/src/main/java/com/messageflow/function/SmsFunction.java
package com.messageflow.function;

import com.google.gson.Gson;
import com.microsoft.azure.functions.*;
import com.microsoft.azure.functions.annotation.AuthorizationLevel;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.annotation.HttpTrigger;
import com.messageflow.function.models.SmsRequest;
import com.messageflow.function.models.SmsResponse;
import com.messageflow.function.services.SmsService;
import com.messageflow.function.utils.Config;

import java.util.Optional;

/**
 * Azure Functions with HTTP Trigger for sending SMS via MessageFlow API
 */
public class SmsFunction {
    private static SmsService smsService;
    private static final Gson gson = new Gson();

    static {
        try {
            Config config = Config.loadConfig();
            smsService = new SmsService(config);
        } catch (Exception e) {
            System.err.println("Failed to initialize SMS service: " + e.getMessage());
        }
    }

    /**
     * This function listens at endpoint "/api/MessageFlowSmsHttpTrigger".
     */
    @FunctionName("MessageFlowSmsHttpTrigger")
    public HttpResponseMessage run(
            @HttpTrigger(
                    name = "req",
                    methods = {HttpMethod.POST},
                    authLevel = AuthorizationLevel.ANONYMOUS)
            HttpRequestMessage<Optional<String>> request,
            final ExecutionContext context) {

        context.getLogger().info("MessageFlow SMS HTTP trigger function processed a request.");

        if (smsService == null) {
            SmsResponse errorResponse = new SmsResponse(
                    false,
                    "SMS service is not available. Please check configuration.",
                    ""
            );
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(errorResponse))
                    .build();
        }

        try {
            // Get JSON data from request
            String body = request.getBody().orElse(null);

            if (body == null || body.isEmpty()) {
                SmsResponse errorResponse = new SmsResponse(
                        false,
                        "Request body is empty",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            context.getLogger().info("Received SMS request data: " + body);

            // Parse request
            SmsRequest smsRequest = gson.fromJson(body, SmsRequest.class);

            // Validate required fields
            if (smsRequest.getSender() == null || smsRequest.getSender().isEmpty()) {
                SmsResponse errorResponse = new SmsResponse(
                        false,
                        "Missing required field: sender",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (smsRequest.getMessage() == null || smsRequest.getMessage().isEmpty()) {
                SmsResponse errorResponse = new SmsResponse(
                        false,
                        "Missing required field: message",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            if (smsRequest.getPhoneNumbers() == null || smsRequest.getPhoneNumbers().isEmpty()) {
                SmsResponse errorResponse = new SmsResponse(
                        false,
                        "Missing required field: phoneNumbers (must be a non-empty array)",
                        ""
                );
                return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "application/json")
                        .body(gson.toJson(errorResponse))
                        .build();
            }

            context.getLogger().info("Created SMS request: sender=" + smsRequest.getSender() +
                    ", message=" + smsRequest.getMessage() +
                    ", phoneNumbers=" + String.join(", ", smsRequest.getPhoneNumbers()));

            // Send SMS
            SmsResponse response = smsService.sendSms(smsRequest);

            HttpStatus status = response.isSuccess() ? HttpStatus.OK : HttpStatus.INTERNAL_SERVER_ERROR;

            return request.createResponseBuilder(status)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(response))
                    .build();

        } catch (Exception e) {
            context.getLogger().severe("Error processing SMS request: " + e.getMessage());
            SmsResponse errorResponse = new SmsResponse(
                    false,
                    "An error occurred: " + e.getMessage(),
                    ""
            );
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                    .header("Content-Type", "application/json")
                    .body(gson.toJson(errorResponse))
                    .build();
        }
    }
}

File: Send-Email-SMS-Azure-Function-MessageFlow-NodeJS/MessageFlowSmsFunction/src/functions/smsFunction.js
const { app } = require('@azure/functions');
const SmsService = require('../services/smsService');

// Create SmsService instance (reused across invocations)
const smsService = new SmsService();

/**
 * Azure Function for sending SMS via MessageFlow API
 * HTTP POST endpoint that accepts SMS data and sends it via MessageFlow
 */
app.http('MessageFlowSmsHttpTrigger', {
  methods: ['POST'],
  authLevel: 'anonymous',
  handler: async (request, context) => {
    context.log('HTTP trigger function processed a SMS request.');

    try {
      // Parse request body
      const requestBody = await request.text();

      if (!requestBody || requestBody.trim() === '') {
        return {
          status: 400,
          body: JSON.stringify({ error: 'Request body is empty' }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      // Parse JSON
      let smsRequest;
      try {
        smsRequest = JSON.parse(requestBody);
      } catch (parseError) {
        context.error('Failed to parse request JSON:', parseError);
        return {
          status: 400,
          body: JSON.stringify({ error: 'Invalid JSON format in request body' }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      // Validate required fields
      if (!smsRequest.sender ||
          !smsRequest.message ||
          !smsRequest.phoneNumbers ||
          !Array.isArray(smsRequest.phoneNumbers) ||
          smsRequest.phoneNumbers.length === 0) {
        return {
          status: 400,
          body: JSON.stringify({
            error: 'Missing required fields: sender, message, and phoneNumbers are required'
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      }

      context.log(`Sending SMS to ${smsRequest.phoneNumbers.length} numbers from ${smsRequest.sender}`);

      // Send SMS using SmsService
      const result = await smsService.sendSmsAsync(smsRequest, context);

      if (result.success) {
        context.log('SMS sent successfully:', result.message);
        return {
          status: 200,
          body: JSON.stringify({
            success: true,
            message: result.message,
            responseContent: result.responseContent
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      } else {
        context.error('Failed to send SMS:', result.message);
        return {
          status: 500,
          body: JSON.stringify({
            success: false,
            message: result.message,
            responseContent: result.responseContent
          }),
          headers: { 'Content-Type': 'application/json' }
        };
      }
    } catch (error) {
      context.error('Unexpected error occurred while processing SMS request:', error);
      return {
        status: 500,
        body: JSON.stringify({
          success: false,
          message: 'An unexpected error occurred',
          error: error.message
        }),
        headers: { 'Content-Type': 'application/json' }
      };
    }
  }
});
